src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,225 @@
+ * any Gamepads and provide a polling system that all devices
+ * will use.  This class assumes that there will only be one
+ * pointer device and one keyboard.
+ */
+export class DeviceInputSystem {
+    // Private Members
+    private _inputs: Map<string, Array<number>> = new Map();
+    private _onDeviceConnected: (deviceName: string) => void = () => { };
+    private _onDeviceDisconnected: (deviceName: string) => void = () => { };
+    private _keyboardActive: boolean = false;
+    private _pointerActive: boolean = false;
+    private _elementToAttachTo: Nullable<HTMLElement>;
+    private _maxKeyCodes : number = 222;
+    private _maxPointerButtons : number = 5;
+
+    /**
+     * Default Constructor
+     * @param elementToAttachTo - element to attach events to (usually canvas)
+     */
+    constructor(elementToAttachTo: Nullable<HTMLElement>) {
+        this._elementToAttachTo = elementToAttachTo;
+        this.handleKeyActions();
+        this.handlePointerActions();
+        this.handleGamepadActions();
+        this.updateDevices();
+    }
+
+    // Public functions
+    /**
+     * Checks for current device input value, given an id and input index
+     * @param deviceName Id of connected device
+     * @param inputIndex Index of device input
+     * @returns Current value of input
+     */
+    public pollInput(deviceName: string, inputIndex: number): number {
+        if (this._inputs.has(deviceName)) {
+            var device = this._inputs.get(deviceName);
+
+            if (device![inputIndex] != undefined) {
+                return device![inputIndex];
+            }
+            else {
+                throw `Unable to find input ${inputIndex} on device ${deviceName}`;
+            }
+        }
+        else {
+            throw `Unable to find device ${deviceName}`;
+        }
+    }
+
+    /**
+     * When a device is connected, perform user specified function
+     * @param callback Callback function to use when a device is connected
+     */
+    public onDeviceConnected(callback: (deviceName: string) => void): void {
+        this._onDeviceConnected = callback;
+    }
+
+    /**
+     * When a device is disconnected, perform user specified function
+     * @param callback Callback function to use when a device is disconnected
+     */
+    public onDeviceDisconnected(callback: (deviceName: string) => void): void {
+        this._onDeviceDisconnected = callback;
+    }
+
+    // Private functions
+    /**
+     * Add device and inputs to device map
+     * @param deviceName Assigned name of device (may be SN)
+     * @param numberOfInputs Number of input entries to create for given device
+     */
+    private registerDevice(deviceName: string, numberOfInputs: number) {
+        if (!this._inputs.has(deviceName)) {
+            var device: Array<number> = [];
+            for (var i = 0; i < numberOfInputs; i++) {
+                device.push(0);
+            }

```

Comment:
I'm initializing the values to zero because the pollInput function checks for undefined as a way of making sure that it's staying within the bounds.  The updateDevice function comes after that. 
 Because of this check, any values that start as undefined will remain undefined.

Reply:
I think this will have subtle bugs if you initialize to 0. If the underlying implementation is event based and the event hasn't fired yet, then polling for a particular input state maybe return 0 when it isn't 0.

Reply:
Originally, I was initializing the values to zero because of how pollInput was checking things.  I found that it would always throw for new values but I've since changed the order of operations.  First, pollInput will check for the device.  If that device exists, it will attempt to update the value for the given index.  Then it'll check if that input is defined.  Finally, it'll return the value.  With this approach, I feel like the second check is kinda unnecessary but it would allow use to not have to initialize values to zero.